### PR TITLE
adding navigation extension types to dynamic-plugin-sdk

### PR DESCRIPTION
**Pull request by glekner**

### CODE REVIEW

**Comment by glekner** [comment]
@vojtechszocs @christianvogt can you review

File: frontend/packages/console-dynamic-plugin-sdk/src/extensions/navigation.ts [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+    /** A unique identifier for this item. */
+    id: string;
+    /** The perspective ID to which this item belongs to. If not specified, contributes to the default perspective. */
+    perspective?: string;
+    /** Navigation section to which this item belongs to. If not specified, render this item as a top level link. */
+    section?: string;
+    /** Mark this item as active when the URL starts with one of these paths. */
+    startsWith?: string[];
+    /** Insert this item before the item referenced here. */
+    insertBefore?: string;
```
Comment (christianvogt) [inline frontend/packages/console-dynamic-plugin-sdk/src/extensions/navigation.ts:14]: We should align with the new capabilities for `insertBefore` and `insertAfter`
```suggestion
    /** Insert this item before the item referenced here. For arrays, the first one found in order is used. */
    insertBefore?: string | string[];
    /** Insert this item after the item referenced here. For arrays, the first one found in order is used. `insertBefore` takes precedence. */
    insertAfter?: string | string[];
```

File: frontend/packages/console-dynamic-plugin-sdk/src/extensions/navigation.ts [hunk 2]
```diff
@@ -0,0 +1,69 @@
+import { Extension } from '@console/plugin-sdk/src/typings/base';
+
+export namespace ExtensionProperties {
```
Comment (christianvogt) [inline frontend/packages/console-dynamic-plugin-sdk/src/extensions/navigation.ts:3]: Can you add `NavSection` as well. This was a recent addition in order to

`console.navigation/section`
```
  export interface NavSection {
    /** A unique identifier for this item. */
    id: string;
    /** The perspective ID to which this item belongs to. If not specified, contributes to the default perspective. */
    perspective?: string;
    /** Name of this section. If not supplied, only a separator will be shown above the section. */
    name?: string;
    /** Insert this item before the item referenced here. For arrays, the first one found in order is used. */
    insertBefore?: string | string[];
    /** Insert this item after the item referenced here. For arrays, the first one found in order is used. `insertBefore` takes precedence. */
    insertAfter?: string | string[];
  }
```

File: frontend/packages/console-dynamic-plugin-sdk/src/extensions/navigation.ts [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+  export type NavItem = {
```
Comment (christianvogt) [inline frontend/packages/console-dynamic-plugin-sdk/src/extensions/navigation.ts:4]: We need a way to supply `data-*` properties to nav items.
Perhaps add `dataAttributes?: {[key: string]: string}`. Where each attribute gets the prefix `data-`
eg.
```
dataAttributes: {
  ['test-id']: 'my-id',
}
```

Results in: `<NavItem data-test-id="my-id"`

```
    /** Adds `data-` attributes to the DOM. Each key will receive the `data-` prefix.  */
    dataAttributes?: {[key: string]: string},
```

**Comment by glekner** [comment]
@christianvogt applied suggestions

**Comment by christianvogt** [comment]
@glekner needs rebase

**Comment by glekner** [comment]
rebased @spadgett @christianvogt

File: frontend/packages/console-dynamic-plugin-sdk/src/extensions/navigation.ts [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+// Type guards
+
+export const isSeparator = (e: Extension): e is Separator =>
+  e.type === 'console.navigation/separator';
+export const isHrefNavItem = (e: Extension): e is HrefNavItem =>
+  e.type === 'console.navigation/href';
+export const isResourceNSNavItem = (e: Extension): e is ResourceNSNavItem =>
+  e.type === 'console.navigation/resource-ns';
+export const isResourceClusterNavItem = (e: Extension): e is ResourceClusterNavItem =>
+  e.type === 'console.navigation/resource-cluster-ns';
```
Comment (christianvogt) [inline frontend/packages/console-dynamic-plugin-sdk/src/extensions/navigation.ts:80]: cluster resource shouldn't include namespace
```suggestion
  e.type === 'console.navigation/resource-cluster';
```

**Comment by glekner** [comment]
fixed @christianvogt

**Pull request merged**